src/include/primer/trie.h

```diff
@@ -0,0 +1,124 @@
+#pragma once
+
+#include <algorithm>
+#include <cstddef>
+#include <future>  // NOLINT
+#include <map>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace bustub {
+
+/// A special type that will block the move constructor and move assignment operator. Used in TrieStore tests.
+class MoveBlocked {
+ public:
+  explicit MoveBlocked(std::future<int> wait) : wait_(std::move(wait)) {}
+
+  MoveBlocked(const MoveBlocked &) = delete;
+  MoveBlocked(MoveBlocked &&that) noexcept {
+    if (!that.waited_) {
+      that.wait_.get();
+    }
+    that.waited_ = waited_ = true;
+  }
+
+  auto operator=(const MoveBlocked &) -> MoveBlocked & = delete;
+  auto operator=(MoveBlocked &&that) noexcept -> MoveBlocked & {
+    if (!that.waited_) {
+      that.wait_.get();
+    }
+    that.waited_ = waited_ = true;
+    return *this;
+  }
+
+  bool waited_{false};
+  std::future<int> wait_;
+};
+
+// A TrieNode is a node in a Trie.
+class TrieNode {
+ public:
+  // Create a TrieNode with no children.
+  TrieNode() = default;
+
+  // Create a TrieNode with some children.
+  explicit TrieNode(std::map<char, std::shared_ptr<const TrieNode>> children) : children_(std::move(children)) {}
+
+  virtual ~TrieNode() = default;
+
+  // Clone returns a copy of this TrieNode. If the TrieNode has a value, the value is copied. The return
+  // type of this function is a shared_ptr to a TrieNode.
```

Comment:
```suggestion
  // type of this function is a unique_ptr to a TrieNode.
```
Did you mean to write unique_ptr?

Reply:
Yes, good spot!